### PR TITLE
6884 viz by name queries 2

### PR DESCRIPTION
**Pull request by javitonino**

### CODE REVIEW

File: app/controllers/visualizations_controller_helper.rb [hunk 1]
```diff
@@ -1,4 +1,56 @@
+require 'carto/uuidhelper'
+
 module VisualizationsControllerHelper
+  class VisualizationLocator
+    extend Carto::UUIDHelper
+
+    def self.parse_table_locator(table_locator)
```
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Introducing `table_locator` to mean visualization "id", "name", "schema.id" or "schema.name".

File: app/controllers/visualizations_controller_helper.rb [hunk 2]
```diff
@@ -1,4 +1,56 @@
...
+    def self.load_user_or_organization(user_or_org_name)
+      user = Carto::User.where(username: user_or_org_name).first
+      if user
+        @user = user
+      else
+        @organization = Carto::organization.where(username: user_or_org_name).first
+      end
+    end
+
+    def self.find(table_locator, user_or_org_name)
```
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Single point of entry to `VisualizationLocator`. The rest of methods will be made private.

This takes the viz_locator and user/org name (a.k.a: subdomain) and returns a visualization or `nil`.

File: app/controllers/visualizations_controller_helper.rb [hunk 3]
```diff
@@ -1,4 +1,56 @@
...
+      elsif @name && @user
+        visualization = get_priority_visualization(@name, user_id: @user.id)
+      elsif @name && @organization && @schema
+        user_from_schema = @organization.users.where(username: @schema).first
+        if user_from_schema
+          visualization = get_priority_visualization(@name, user_id: user_from_schema.id)
+        end
+      end
+
+      visualization if visualization && validate_result(visualization)
```
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Enforced validation. Only returns visualizations that match all the search parameters given.

File: app/controllers/visualizations_controller_helper.rb [hunk 4]
```diff
@@ -1,4 +1,56 @@
...
+    end
+
+    def self.validate_result(visualization)
```
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: This does not validate URL's like: `org.cartodb.com/u/user` where `user` is not part of `org`. I think this should be the concern of another helper, based on
```
  def eligible_for_redirect?(user)
    return false if CartoDB.subdomainless_urls?
    user.has_organization? && !request.params[:redirected].present? &&
      CartoDB.subdomain_from_request(request) != user.organization.name
  end
```

File: app/controllers/visualizations_controller_helper.rb (hunk 1 above)
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: I'm considering dropping the schema entirely for new URLs. It does not add new information and may generate conflicts when it differs from the username in the url.

File: app/controllers/visualizations_controller_helper.rb [hunk 5]
```diff
@@ -7,20 +59,15 @@ def get_priority_visualization(visualization_id, user_id: nil, organization_id:
...
                                     .all
-                                    .sort do |vis_a, _vis_b|
+                                    .sort { |vis_a, _vis_b|
                                       vis_a.type == Carto::Visualization::TYPE_CANONICAL ? -1 : 1
-                                    end
+                                    }
                                     .first
   end
 
   def load_visualization_from_id_or_name(id_or_name)
```
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: We would call this from most controllers.

File: app/controllers/visualizations_controller_helper.rb [hunk 6]
```diff
@@ -1,4 +1,56 @@
+require 'carto/uuidhelper'
+
 module VisualizationsControllerHelper
+  class VisualizationLocator
```
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Extracted as a class, although it could be a (massive) methods. I think this makes this easier to read and test.

**Comment by javitonino** [comment]
TLDR: Finding visualizations is surprisingly complex. I suggest:
- New code: Do not allow schema in table names. Enforce `/u/user` as part of the URL for org users. Avoid the code in this PR.
- Old code: Minimal changes. Fix what's broken.

---

See commit: https://github.com/CartoDB/cartodb/pull/6921/commits/e50b3b4ca07374621ff5280b072aea9f4a864ffd

This is a prototype for a visualization locator, in order to extract the behaviour to find a given visualization by id/name and username/org. It is quite complex and only handles valid (full information available) cases.

However, while writing all of this, the following has come to mind, we can simplify a lot of this code by dropping the possibility of adding the schema for new urls. We would enforce that all urls specify an user, i.e: `/u/username` is required for organization urls. This only leaves to parameters: either viz id or name, and username and makes all of this much nicer.

For old code, I'm thinking of leaving all as it is, except for problematic parts. There is an interesting problem for URL's of the form `org.cartodb.com/tables/name`. The frontend extracts the parameters from the URL for the SQL-API call, and uses `org` as an username. So, those URL's do not load data from the table (they show the map correctly as that uses the viz_id from the vizjson). I would redirect this URL's to the correct ones.

cc @juanignaciosl @gfiorav

**Comment by juanignaciosl** [comment]
I'm not sure that dropping the schema from the table name is a change that we can do, I think we must support it. What we can do is simplifying it. Detecting it is easy (split with `.`, discard `public`), and always take precedence over username, for example.

That being said, I'll take look at the code.

File: app/controllers/visualizations_controller_helper.rb [hunk 7]
```diff
@@ -1,4 +1,56 @@
...
+    extend Carto::UUIDHelper
+
+    def self.parse_table_locator(table_locator)
+      table_id_or_name, @schema = table_locator.split('.').reverse
+
+      if is_uuid?(table_id_or_name)
+        @id = table_id_or_name
```
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: This is setting the `@id` in the _class_, since `parse_table_locator` is a class method. I guess this is not intended, isn't it?

Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Many design changes along the way and the final version is so screwed up, You're right, of course!

File: app/controllers/visualizations_controller_helper.rb (hunk 6 above)
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: I've taken a first look, and my feeling is that this has become too complicated to manage, specially because it handles a lot of state (`@id`, `@user`, `@organization`. `@name`...), which skyrockets the casuistic. Quick idea: couldn't this be split into two concerns, parsing parameters and searching? Instead of thinking about a `VisualizationLocator` responsible of both, maybe there could be a `VisualizationIdentification` structure generated by a factory method responsible of parsing the fuzzy inputs ("ids or names", "usernames or organization names") into well known attributes ("username", "organization name"...). Then we could add another factory method to `VisualizationQueryBuilder` so we could use the `VisualizationIdentification` in an easy way.

I'll wait for your opinion on this before continuing. Maybe you already tried and you know this is a worse approach.

Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: I agree with the complexity, that's why I though of dropping some parsing in new URL's. I think that for strict URL validation (search only by uuid, or user+viz_name) I can make a decent split. Let me see what I can come up with.

**Comment by javitonino** [comment]
I think this is better:
https://github.com/CartoDB/cartodb/blob/6884-viz_by_name_queries_2/app/controllers/visualizations_controller_helper.rb

There are three components to this:
- `VisualizationLocator` that stores and parses the visualization name (which may include the schema).
- `extract_user_from_request_and_viz_locator` which returns the owner of the visualization.
- `load_visualization_from_id_or_name` that recovers the visualization itself.

Take a look @juanignaciosl

File: app/controllers/visualizations_controller_helper.rb [hunk 8]
```diff
@@ -1,24 +1,90 @@
...
+      table_id_or_name, @schema = visualization_locator.split('.').reverse
+
+      if is_uuid?(table_id_or_name)
+        @id = table_id_or_name
+      else
+        @name = table_id_or_name
+      end
+    end
+
+    def id?
```
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: I may have gone overboard with these. It makes the following code a little easier to read, but I may delete them.

Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: Yep, definitely.

File: app/controllers/visualizations_controller_helper.rb [hunk 9]
```diff
@@ -1,24 +1,90 @@
+require_dependency 'carto/uuidhelper'
+
 module VisualizationsControllerHelper
+  class VisualizationLocator
+    include Carto::UUIDHelper
+
+    def initialize(visualization_locator)
+      table_id_or_name, @schema = visualization_locator.split('.').reverse
```
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: `.reverse` + multiple assignment is a nice and easy trick! I foresee many refactors :-) <3

File: app/controllers/visualizations_controller_helper.rb [hunk 10]
```diff
@@ -1,24 +1,90 @@
...
+
+      if is_uuid?(table_id_or_name)
```
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: If a table has a UUID name this won't find it. I'm not sure wether we support it right now and there's no such value in production, but... :sweat_smile:

File: app/controllers/visualizations_controller_helper.rb [hunk 11]
```diff
@@ -1,24 +1,90 @@
...
+      !@schema.nil?
+    end
+
+    def schema
+      @schema
+    end
+
+    def matches_visualization?(visualization)
+      return false unless visualization
+      return false if id? && id != visualization.id
```
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: I think you don't have to do nil checks here, comparing a value to nil is safe.

In case that you're trying to return `false` if the value has not been set this is probably wrong, because some of them can be `nil` but still be a match.

File: app/controllers/visualizations_controller_helper.rb [hunk 12]
```diff
@@ -1,24 +1,90 @@
...
+      return false if name? && name != visualization.name
+      return false if schema? && schema != visualization.user.database_schema
+      true
+    end
+  end
+
+  def extract_user_from_request_and_viz_locator(viz_locator)
```
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: I guess this is because of the design iterations, you probably don't want a `viz_locator` parameter, since this is an instance method.

File: app/controllers/visualizations_controller_helper.rb (hunk 11 above)
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: I need the check because if the value has not been set (is nil), it does not restricts what can match. 
Example: If I only pass the visualization name in the constructor, `id` and `schema` will be `nil` so I can't see if they match. So I return true.

This check basically only makes sense for the schema, as it can be ignored in certain situation, so this is my way of making sure it is consistent with the retrieved visualization. Either `id` or `name` will always be present and in the code below, I search using them, so unless the db breaks horribly, it will match.

File: app/controllers/visualizations_controller_helper.rb (hunk 10 above)
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: We are safe, no hyphens allowed in the table name, they are converted to underscores.

File: app/controllers/visualizations_controller_helper.rb (hunk 12 above)
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: This is outside the VizLocator class (which is limited to parsing and validating the `schema.table` thingy).

File: app/controllers/visualizations_controller_helper.rb [hunk 13]
```diff
@@ -1,24 +1,90 @@
...
+    # 1. Handles any url with "/u/username", or "username.cartodb.com"
+    user_or_org_name = CartoDB.extract_subdomain(request)
+    user = Carto::User.where(username: user_or_org_name).first
+    return user unless user.nil?
+
+    # 2. Handles org.cartodb.com with "schema.table" visualizations
+    if viz_locator.schema?
+      organization = Carto::Organization.where(name: user_or_org_name).first
+      return nil unless organization
+      organization.users.where(username: viz_locator.schema).first
```
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: Since `schema` might be `public` instead of a direct `schema` usage you might need here a

``ruby
def username_from_schema
  @schema == 'public' ? nil : @schema
end
```

File: app/controllers/visualizations_controller_helper.rb (hunk 12 above)
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: Oh, now I see it's a module method, my bad.

File: app/controllers/visualizations_controller_helper.rb (hunk 13 above)
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Well, either `schema` is an user from the organization, and this returns an user, or it is whatever (including `public`) and this returns `nil`. This explicitly does not returns users if they are not well specified (by url parameters, domain, or schema).

Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Checking for `public` saves a trip to the database in some cases (given organization, not username, and schema=public), but that is a rare case and I prefer the simplicity of this.

File: app/controllers/visualizations_controller_helper.rb (hunk 10 above)
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: I must disagree, there're several visualizations matching `'%-%-%-%-%'`.

File: app/controllers/visualizations_controller_helper.rb (hunk 11 above)
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: :ok_hand:

File: app/controllers/visualizations_controller_helper.rb (hunk 10 above)
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Wow, major oversight by me, was looking only at table visualizations. Thanks for bringing it up. This is already a problem for existing code then, as `VQB::with_id_or_name` does the same thing.

I will add a fallback here, so that if uuid search fails, it searches by name and user. I will also open an issue to fix the problem elsewhere.

Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Uhmm, maybe it is not a problem, as table visualizations can't be uuid and are the ones being search by name.

I will look into the code to see if this may trigger in some instances and add some tests, because it is something we must keep in mind.

Q: Is the editor-3 supposed to work with canonical visualizations?

Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: You can probably defer the "uuid name" issue, 100% not relevant right now.

Q: we don't know yet, but probably yes.

**Comment by javitonino** [comment]
I made `carto/api/visualizations_controller.rb` to call the new visualization loader. It was reported by Rollbar as creating long queries and was a duplication of the `get_priority_visualization` code.

**Comment by javitonino** [comment]
I merged master (easier to read the changes now) and moved the 404 rendering outside of the helper, to the controllers, as it should be their concern. Aside from reorganizing the visualization loading code, this fixes the last couple of slow queries remaining.

Tests are green, they don't show on Github, probably because their downtime yesterday; so I think this is ready. Please take a look @juanignaciosl (I don't remember exactly where we left this).

File: app/controllers/visualizations_controller_helper.rb [hunk 14]
```diff
@@ -1,6 +1,55 @@
+require_dependency 'carto/uuidhelper'
 require_dependency 'carto/api/vizjson3_presenter'
 
 module VisualizationsControllerHelper
+  class VisualizationLocator
+    include Carto::UUIDHelper
+
+    def initialize(visualization_locator)
```
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: Could you please rename `visualization_locator` to `visualization_locator_string` (or a better name if you come up with any) and document in the constructor what inputs does it accept (schema.name, name, id...). A comment about the motivation ("accept flexible visualization identification" or something similar might be interesting as well. Code is clear but IMHO this requires a little of documentation.

File: app/controllers/visualizations_controller_helper.rb [hunk 15]
```diff
@@ -1,6 +1,55 @@
...
+
+    def schema
+      @schema
+    end
+
+    def matches_visualization?(visualization)
+      return false unless visualization
+      return false if id && id != visualization.id
+      return false if name && name != visualization.name
+      return false if schema && schema != visualization.user.database_schema
```
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: I think this is ok, but keep an eye on organization complaints, since some might be using a schema of another user through a share. Could you test in staging the current behaviour of using the schema of a user which has the shared dataset or map instead of the dataset / map owner, please?

**Comment by juanignaciosl** [comment]
LGTM. I've done a couple of minor comments, please take into account the extra test suggested for staging.

File: app/controllers/visualizations_controller_helper.rb (hunk 15 above)
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: This was OK, but the user finding code was wrong. The tests caught it. I have modified `extract_user_from_request_and_viz_locator` to correctly handle this case (logged in user accessing another user's viz).

**Comment by javitonino** [comment]
I added "some" tests and caught a bug (related to shared visualizations).

File: spec/requests/visualizations_controller_helper_spec.rb [hunk 16]
```diff
@@ -0,0 +1,321 @@
+# encoding: utf-8
+require_relative '../spec_helper'
```
Comment (javitonino) [inline spec/requests/visualizations_controller_helper_spec.rb]: This is not in `spec_helper_min` as I could not manage to create an organization with that helper. The existing factories create the old models and need the full spec_helper, and I was not able to create a factory for the new models (some problems with saving the whitelisted email list).
